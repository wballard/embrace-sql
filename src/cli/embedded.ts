import { Command } from "commander";
import { loadConfiguration } from "../configuration";
import { buildInternalContext, InternalContext } from "../context";
import { watchRoot } from "../watcher";
import expandHomeDir from "expand-home-dir";
import path from "path";
import ncc from "@zeit/ncc";
import fs from "fs-extra";

/**
 * Command line option type for the typescript!
 */
type options = {
  nowatch: boolean;
};

/**
 * Bundle/compile embraceSQL itself into a single emebdded module.
 *
 * This doesn't bundle the handlers or generated code, but the core engine.
 * The idea here is to have no opportunity to have a version mismatch of
 * EmbraceSQL itself -- the CLI you use to generate is the CLI used to create
 * the runtime. Its - sorta like pulling itsef up by it's sql-y bootstraps
 */
export const bundleEmbedded = async (
  EMBRACESQL_ROOT: string
): Promise<void> => {
  try {
    const sourceFile = path.join(__dirname, "..", "embedded.ts");
    const { code, assets } = await ncc(sourceFile, {
      // provide a custom cache path or disable caching
      cache: path.resolve(path.join(EMBRACESQL_ROOT, ".cache")),
      // externals to leave as requires of the build
      externals: [],
      // directory outside of which never to emit assets
      filterAssetBase: EMBRACESQL_ROOT,
      minify: false,
      sourceMap: true,
      sourceMapBasePrefix: EMBRACESQL_ROOT,
      sourceMapRegister: true,
      watch: false,
      v8cache: false,
      quiet: false,
      debugLog: false,
    });
    await fs.outputFile(path.join(EMBRACESQL_ROOT, "index.js"), code);
    // type definitions are found in the assets
    const waitForAssets = Object.keys(assets).map((assetFileName) => {
      if (assetFileName.endsWith("embedded.d.ts")) {
        const assetContent = (assets[assetFileName].source as Buffer).toString(
          "utf-8"
        );
        return fs.outputFile(
          path.join(EMBRACESQL_ROOT, "embedded.d.ts"),
          assetContent
        );
      }
    });
    await Promise.all(waitForAssets);
  } catch (e) {
    console.error(e);
  } finally {
  }
};

/**
 * Initialization action.
 */
export default new Command()
  .command("embedded [EMBRACEQL_ROOT]")
  .option("-n, --nowatch", "Disable watching for changes")
  .description("Generate an embebbed version of EmbraceSQL.")

  .action(
    async (EMBRACEQL_ROOT: string, cmd: options): Promise<void> => {
      // fully qualified path from here on down will make things a lot simpler
      const root = path.resolve(
        expandHomeDir(
          EMBRACEQL_ROOT || process.env.EMBRACEQL_ROOT || "/var/embracesql"
        )
      );
      const configuration = await loadConfiguration(root, true);
      const internalContext = await buildInternalContext(configuration);
      // just needed for generation -- so close it off.
      internalContext.close();
      // no watch -- bail
      if (cmd.nowatch) return;

      const watcher = watchRoot(root);
      watcher.emitter.on("reload", async (newContext: InternalContext) => {
        console.info("Reloading");
        // just needed for generation -- so close it off.
        await newContext.close();
      });
    }
  );
