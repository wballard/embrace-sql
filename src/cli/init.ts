import { Command } from "commander";
import { generateFromTemplates } from "../generator";
/**
 * Initialization action.
 */
export const init = new Command()
  .command("init")
  .description("Generates a docker-compose.yaml so you can get started.")
  .action(async () => {
    await generateFromTemplates(undefined, "cli", true);
  });
