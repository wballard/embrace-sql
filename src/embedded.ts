import { InternalContext, buildInternalContext } from "./context";
import { loadConfiguration } from "./configuration";

/**
 * Given a root directory, return an embedded context ready to be used with
 * and inprocess client.
 */
export const buildEmbeddedContext = async (
  root: string
): Promise<InternalContext> => {
  const configuration = await loadConfiguration(root);
  return buildInternalContext(configuration);
};
