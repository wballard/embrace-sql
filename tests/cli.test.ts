import init from "../src/cli/init";
import embedded from "../src/cli/embedded";
import sinon from "sinon";
import path from "path";
import readFile from "read-file-utf8";

describe("CLI", () => {
  const sandbox = sinon.createSandbox();
  beforeEach(() => {});
  afterEach(() => {
    sandbox.restore();
  });
  it("initializes", async () => {
    const spy = sandbox.spy(process.stdout, "write");
    const command = await init.parseAsync(["node", "init"]);
    // trick needed to wait out
    await Promise.all(command.parent?._actionResults || []);
    expect(command.args).toMatchSnapshot();
    const output = spy.getCall(0).args[0];
    expect(output).toMatchSnapshot();
  });
  it("generates an embedded client", async () => {
    const root = path.resolve(path.join(__dirname, "..", ".tests", "embedded"));
    const command = await embedded.parseAsync(["node", "embedded", root, "--nowatch"]);
    // trick needed to wait out
    await Promise.all(command.parent?._actionResults || []);
    expect(command.args).toMatchSnapshot();
    // check on the key files
    expect(await readFile(path.join(root, "index.ts"))).toMatchSnapshot();
  });
});
