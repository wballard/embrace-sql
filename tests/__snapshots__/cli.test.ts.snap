// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CLI generates an embedded client 1`] = `
Array [
  "/Users/wballard/embrace-sql-core/.tests/embedded",
]
`;

exports[`CLI generates an embedded client 2`] = `
"import { buildEmbeddedContext } from \\"embracesql\\";
import { EmbraceSQL } from \\"./client/node-inprocess\\";

/**
 * Use this to access the database.
 */
export const EmbraceSQLEmbedded = async () => {
  // the package itself has the configuration and is the root
  const context = await buildEmbeddedContext(__dirname);
  return EmbraceSQL(context);
};
"
`;

exports[`CLI initializes 1`] = `Array []`;

exports[`CLI initializes 2`] = `
"version: '3.0'
services:
  embracesql:
    command: start /var/embracesql 8765
    image: embracesql/server
    ports:
    - \\"8765:8765\\"
    volumes:
    - .:/var/embracesql"
`;
